File: ruby/chapter_1/movie.rb
require_relative './price'

class Movie
  @CHILDRENS = 2
  @REGULAR = 0
  @NEW_RELEASE = 1

  class << self
    attr_accessor :CHILDRENS, :REGULAR, :NEW_RELEASE
  end

  attr_reader :title, :price

  def initialize(title, code)
    @title = title
    set_price(code)
  end

  def price_code
    price.code
  end

  def set_price(code)
    @price = case code
            when Movie.CHILDRENS
              ChildrenPrice
            when Movie.NEW_RELEASE
              NewReleasePrice
            when Movie.REGULAR
              RegularPrice
            end.new
  end

  def charge(days_rented)
    price.get_charge(days_rented)
  end

  def points(days_rented)
    price.get_points(days_rented)
  end
end

class Rental
  attr_reader :movie, :days_rented
  def initialize(movie, days_rented)
    @movie = movie
    @days_rented = days_rented
  end

  def get_charge
    movie.charge(days_rented)
  end

  def get_points
    movie.points(
      days_rented)
  end
end

class Customer
  attr_reader :name, :rentals

  def initialize(name)
    @name = name
    @rentals = []
  end

  def add_rental(rental)
    rentals << rental
  end

  def statement
    result = "Rental Record for #{name} \n"

    rentals.each do |each|    
      result += "\t#{each.movie.title}\t#{each.get_charge}\n"
    end

    result += "Amount owed is #{get_total_charge}\n"
    result += "You earned #{get_rental_points} frequent renter points"
    result
  end

  def html_statement
    result = "<h1>Rental Record for <em>#{name}</em></h1>\n"
    
    rentals.each do |each|    
      result += "<p>#{each.movie.title}: #{each.get_charge}</p>\n"
    end

    result + "<p>You owe #{get_total_charge} and earned #{get_rental_points} frequent renter points</p>\n"
  end

  def get_total_charge
    rentals.reduce(0) do |total, rental|
      total + rental.get_charge
    end
  end

  def get_rental_points
    rentals.reduce(0) do |total, rental|
      total + rental.get_points
    end
  end
end

File: javascript/chapter_1/movie.js
// Explicit to this module

let map = new WeakMap()
let internal = function(object) {
  if (!map.has(object)) map.set(object, {})
  return map.get(object)
}

class Movie {
  static get REGULAR() {
    return 0
  }
  static get NEW_RELEASE() {
    return 1
  }
  static get CHILDREN() {
    return 2
  }
  constructor(title, priceCode) {
    internal(this).title = title
    internal(this).price = this.setPriceCode(priceCode)
  }
  getPriceCode() {
    return internal(this).price
  }
  setPriceCode(arg) {
    switch (arg) {
      case Movie.REGULAR:
        return new RegularPrice()
      case Movie.NEW_RELEASE:
        return new NewReleasePrice()
      case Movie.CHILDREN:
        return new ChildrensPrice()
      default:
        throw new Error('Incorrect Price Code')
    }
  }
  getTitle() {
    return internal(this).title
  }
  getCharge(daysRented) {
    return internal(this).price.getCharge(daysRented)
  }

  frequentRenterPoints(daysRented) {
    return internal(this).price.getFrequentRenterPoints(daysRented)
  }
}

class Price {
  constructor() {
    if (this.constructor === Price) {
      throw new Error("Abstract class: can't be directly instantiated")
    }
  }
  getPriceCode() {
    throw new Error('Class instance must overright this method')
  }
  getCharge() {
    throw new Error('Class instance must overright this method')
  }
  getFrequentRenterPoints(daysRented) {
    return 1
  }
}

class ChildrensPrice extends Price {
  getPriceCode() {
    return Movie.CHILDRENS
  }
  getCharge(daysRented) {
    let result = 1.5
    if (daysRented > 3) result += (daysRented - 3) * 1.5
    return result
  }
}
class NewReleasePrice extends Price {
  getPriceCode() {
    return Movie.NEW_RELEASE
  }
  getCharge(daysRented) {
    return daysRented * 3
  }
  getFrequentRenterPoints(daysRented) {
    return daysRented > 1 ? 2 : 1
  }
}
class RegularPrice extends Price {
  getPriceCode() {
    return Movie.REGULAR
  }
  getCharge(daysRented) {
    let result = 2
    if (daysRented > 2) result += (daysRented - 2) * 1.5
    return result
  }
}

module.exports = { Movie, ChildrensPrice, NewReleasePrice, RegularPrice }
